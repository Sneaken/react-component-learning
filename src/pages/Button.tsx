import { Button } from '@/components';
import { AppleOutlined } from '@ant-design/icons';
import { Button as AntdButton, Checkbox, Form, Input, Radio } from 'antd';
import { type HTMLAttributeAnchorTarget } from 'react';

const icon = <AppleOutlined />;

function ButtonOfPage() {
  const [form] = Form.useForm();

  const size = Form.useWatch('size', form);
  const type = Form.useWatch('type', form);
  const shape = Form.useWatch('shape', form);
  const block = Form.useWatch('block', form);
  const danger = Form.useWatch('danger', form);
  const disabled = Form.useWatch('disabled', form);
  const ghost = Form.useWatch('ghost', form);
  const showIcon = Form.useWatch('showIcon', form);
  const children = Form.useWatch('children', form);
  const loading = Form.useWatch('loading', form);
  const href = Form.useWatch('href', form);
  const target: HTMLAttributeAnchorTarget = Form.useWatch('target', form);

  return (
    <div>
      <Form form={form} initialValues={{ size: 'middle' }}>
        <Form.Item label="size" name="size">
          <Radio.Group>
            <Radio.Button value="small">small</Radio.Button>
            <Radio.Button value="middle">middle</Radio.Button>
            <Radio.Button value="large">large</Radio.Button>
          </Radio.Group>
        </Form.Item>
        <Form.Item label="type" name="type">
          <Radio.Group>
            <Radio.Button value="default">default</Radio.Button>
            <Radio.Button value="primary">primary</Radio.Button>
            <Radio.Button value="ghost">ghost</Radio.Button>
            <Radio.Button value="dashed">dashed</Radio.Button>
            <Radio.Button value="link">link</Radio.Button>
            <Radio.Button value="text">text</Radio.Button>
          </Radio.Group>
        </Form.Item>
        <Form.Item label="shape" name="shape">
          <Radio.Group>
            <Radio.Button value="default">default</Radio.Button>
            <Radio.Button value="circle">circle</Radio.Button>
            <Radio.Button value="round">round</Radio.Button>
          </Radio.Group>
        </Form.Item>
        <Form.Item label="target" name="target">
          <Radio.Group>
            <Radio.Button value="_self">_self</Radio.Button>
            <Radio.Button value="_blank">_blank</Radio.Button>
            <Radio.Button value="_parent">_parent</Radio.Button>
            <Radio.Button value="_top">_top</Radio.Button>
          </Radio.Group>
        </Form.Item>
        <Form.Item label="block" name="block" valuePropName="checked">
          <Checkbox />
        </Form.Item>
        <Form.Item label="danger" name="danger" valuePropName="checked">
          <Checkbox />
        </Form.Item>
        <Form.Item label="disabled" name="disabled" valuePropName="checked">
          <Checkbox />
        </Form.Item>
        <Form.Item label="loading" name="loading" valuePropName="checked">
          <Checkbox />
        </Form.Item>
        <Form.Item label="ghost" name="ghost" valuePropName="checked">
          <Checkbox />
        </Form.Item>
        <Form.Item label="showIcon" name="showIcon" valuePropName="checked">
          <Checkbox />
        </Form.Item>
        <Form.Item label="children" name="children">
          <Input />
        </Form.Item>
        <Form.Item label="href" name="href">
          <Input allowClear />
        </Form.Item>
      </Form>
      <div style={{ float: 'left', width: '40vw' }}>
        <Button
          size={size}
          type={type}
          shape={shape}
          block={block}
          danger={danger}
          disabled={disabled}
          ghost={ghost}
          icon={showIcon ? icon : undefined}
          children={children}
          loading={loading}
          href={href}
          target={target}
        />
      </div>
      <div style={{ float: 'left', width: '40vw' }}>
        <AntdButton
          size={size}
          type={type}
          shape={shape}
          block={block}
          danger={danger}
          disabled={disabled}
          ghost={ghost}
          icon={showIcon ? icon : undefined}
          children={children}
          loading={loading}
          href={href}
          target={target}
        />
      </div>
    </div>
  );
}

export default ButtonOfPage;
