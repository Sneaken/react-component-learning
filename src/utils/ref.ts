import { type Ref, useRef } from 'react';
import { isMemo } from 'react-is';

export function fillRef<T>(ref: Ref<T>, node: T) {
  if (typeof ref === 'function') {
    ref(node);
  } else if (typeof ref === 'object' && ref && Object.hasOwn(ref, 'current')) {
    // MutableRefObject
    (ref as any).current = node;
  }
}

/**
 * Merge refs into one ref function to support ref passing.
 */
export function composeRef<T>(...refs: Ref<T>[]): Ref<T> {
  const refList = refs.filter((ref) => ref);
  if (refList.length <= 1) return refList[0];

  return (node: T) => {
    refs.forEach((ref) => {
      fillRef(ref, node);
    });
  };
}

interface Cache<Value, Condition> {
  condition?: Condition;
  value?: Value;
}

export function useComposeRef<T>(...refs: Ref<T>[]): Ref<T> | undefined {
  const cacheRef = useRef<Cache<Ref<T>, Ref<T>[]>>({});
  const shouldUpdate = (prev: Ref<T>[], next: Ref<T>[]) =>
    prev.length === next.length && prev.every((ref, i) => ref === next[i]);

  if (!Object.hasOwn(cacheRef.current, 'value') || shouldUpdate(cacheRef.current.condition!, refs)) {
    cacheRef.current.value = composeRef(...refs);
    cacheRef.current.condition = refs;
  }

  return cacheRef.current.value;
}

export function supportRef(nodeOrComponent: any): boolean {
  const type = isMemo(nodeOrComponent) ? nodeOrComponent.type.type : nodeOrComponent.type;
  // Function component node 不是 forwardRef 包裹的时候是不能挂载 ref 的
  if (typeof type === 'function' && !type.prototype?.render) return false;
  // Class component 没有 render 函数也是一样
  if (typeof nodeOrComponent === 'function' && !nodeOrComponent.prototype?.render) return false;
  return true;
}
