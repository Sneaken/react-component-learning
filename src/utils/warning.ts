import noop from '@/utils/noop';

type Warning = (valid: boolean, component: string, message?: string) => void;

export let warning: Warning = noop;

if (process.env.NODE_ENV !== 'production') {
  warning = (valid, component, message) => {
    if (!valid && console !== undefined) {
      console.error(`Warning: [${component}] ${message}`);
    }
  };
}
