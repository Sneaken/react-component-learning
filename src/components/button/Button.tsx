import Wave from '@/components/wave';
import omit from '@/utils/omit';
import { warning } from '@/utils/warning';
import classnames from 'classnames';
import type {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  MouseEvent as ReactMouseEvent,
  MouseEventHandler,
  ReactNode,
} from 'react';
import { createRef, forwardRef, useEffect, useState } from 'react';
import LoadingIcon from './LoadingIcon';

import './style/index.less';

export type SizeType = 'small' | 'middle' | 'large';

export type ButtonType = 'default' | 'primary' | 'ghost' | 'dashed' | 'link' | 'text';

export type ButtonShape = 'default' | 'circle' | 'round';

export interface BaseButtonProps {
  // 将按钮宽度调整为其父宽度的选项
  block?: boolean;
  danger?: boolean;
  disabled?: boolean;
  // 使按钮背景透明
  ghost?: boolean;
  href?: string;
  htmlType?: string;
  icon?: ReactNode;
  loading?: boolean | { delay?: number };
  onClick?: (event: MouseEvent) => void;
  shape?: ButtonShape;
  size?: SizeType;
  target?: string;
  type?: ButtonType;
}

export type AnchorButtonProps = {
  href: string;
  target?: string;
  onClick?: MouseEventHandler<HTMLAnchorElement>;
} & BaseButtonProps &
  Omit<AnchorHTMLAttributes<any>, 'type' | 'onClick'>;

export type ButtonHTMLType = 'submit' | 'button' | 'reset';

export type NativeButtonProps = {
  htmlType?: ButtonHTMLType;
  onClick?: MouseEventHandler<HTMLButtonElement>;
} & BaseButtonProps &
  Omit<ButtonHTMLAttributes<any>, 'type' | 'onClick'>;

export type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>;

type Loading = number | boolean;

function isUnBorderedButtonType(type: ButtonType | undefined) {
  return type === 'text' || type === 'link';
}

const Button = forwardRef<HTMLButtonElement | HTMLAnchorElement, ButtonProps>(function InternalButton(props, ref) {
  const {
    block = false,
    children,
    className,
    danger,
    disabled,
    ghost,
    htmlType = 'button' as ButtonProps['htmlType'],
    icon,
    loading = false,
    shape = 'default',
    size = 'middle',
    type = 'default',
    ...rest
  } = props;

  const [innerLoading, setLoading] = useState<Loading>(!!loading);
  // 如果外部没绑定 ref
  // 但是为什么用 createRef
  const buttonRef = (ref as any) || createRef<HTMLButtonElement | HTMLAnchorElement>();

  // =============== Update Loading ===============
  const loadingOrDelay: Loading = typeof loading === 'boolean' ? loading : loading?.delay || true;

  useEffect(() => {
    let delayTimer: number | null = null;

    if (typeof loadingOrDelay === 'number') {
      delayTimer = window.setTimeout(() => {
        delayTimer = null;
        setLoading(loadingOrDelay);
      }, loadingOrDelay);
    } else {
      setLoading(loadingOrDelay);
    }

    return () => {
      if (!delayTimer) return;
      window.clearTimeout(delayTimer);
      delayTimer = null;
    };
  }, [loadingOrDelay]);

  const handleClick = (e: ReactMouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>) => {
    const { onClick } = props;
    if (innerLoading || disabled) {
      e.preventDefault();
      return;
    }
    (onClick as MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>)?.(e);
  };

  if (process.env.NODE_ENV !== 'production') {
    switch (size) {
      case 'large':
      case 'small':
      case 'middle':
      case undefined:
        break;
      default:
        warning(!size, 'Button', 'Invalid prop `size`.');
    }
  }

  warning(!(ghost && isUnBorderedButtonType(type)), 'Button', "`link` or `text` button can't be a `ghost` button.");

  // react-router 里面 Link 组件 允许使用自定义 组件 需要透传 navigate （可能是旧版本的react-router-dom 待定）
  const linkButtonRestProps = omit(rest as AnchorButtonProps & { navigate: any }, ['navigate']);

  const hrefAndDisabled = linkButtonRestProps.href !== undefined && disabled;

  const iconType = innerLoading ? 'loading' : icon;

  const _className = classnames(
    'btn',
    {
      'btn-background-ghost': ghost && !isUnBorderedButtonType(type),
      'btn-block': block,
      'btn-dangerous': !!danger,
      'btn-disabled': hrefAndDisabled,
      'btn-icon-only': !children && children !== 0 && iconType,
      'btn-loading': innerLoading,
      // 'btn-rtl': direction === 'rtl',
      [`btn-${shape}`]: shape !== 'default' && shape,
      [`btn-${size}`]: size,
      [`btn-${type}`]: type,
    },
    className
  );

  const iconNode = icon && !innerLoading ? icon : <LoadingIcon existIcon={!!icon} loading={!!innerLoading} />;

  // href="" 也是 有意义的
  if (linkButtonRestProps.href !== undefined) {
    return (
      <a {...linkButtonRestProps} className={_className} onClick={handleClick} ref={buttonRef}>
        {iconNode}
        {children}
      </a>
    );
  }

  let buttonNode = (
    <button {...rest} className={_className} ref={buttonRef} type={htmlType} disabled={disabled} onClick={handleClick}>
      {iconNode}
      {children}
    </button>
  );

  if (!isUnBorderedButtonType(type)) {
    // 点击会出现波浪的动画
    buttonNode = <Wave disabled={!!innerLoading}>{buttonNode}</Wave>;
  }

  return buttonNode;
});

if (process.env.NODE_ENV !== 'production') {
  Button.displayName = 'Button';
}

export default Button;
