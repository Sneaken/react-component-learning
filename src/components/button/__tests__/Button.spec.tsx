import noop from '@/utils/noop';
import { resetWarned } from '@/utils/warning';
import mountTest from '@tests/shared/mountTest';
import { act, fireEvent, render, waitFakeTimer } from '@tests/utils';
import { useState } from 'react';
import Button, { type ButtonProps } from '..';

describe('Button', () => {
  mountTest(Button);
  mountTest(() => <Button size="large" />);
  mountTest(() => <Button size="small" />);

  it('renders correctly', () => {
    const { container } = render(<Button>Follow</Button>);
    // 如果没有快照会重新生成一份
    // 如果存在快照，则跟上次生成的快照进行对比，如果一致测试通过，如果不一致则说明代码有改动。
    // 如果确实要更新快照 则传入 --update 来更新快照
    // vitest --update
    // container.firstChild 说的就是 Button 组件
    expect(container.firstChild).toMatchSnapshot();
  });

  it('mount correctly', () => {
    // 渲染正常没有抛出异常
    expect(() => render(<Button>Follow</Button>)).not.toThrow();
  });

  it('warns if size is wrong', () => {
    resetWarned();
    // vi.spyOn: Creates a spy on a method or getter/setter of an object.
    // mockImplementation: Accepts a function that will be used as an implementation of the mock.
    // 如果不传一个空函数会把错误信息打印的控制台上（从行为上来说这个正确的, 但是从看测试报告的角度来看是不想要看到的）
    const mockWarn = vi.spyOn(console, 'error').mockImplementation(noop);
    const size = 'who am I';
    // @ts-expect-error: Type '"who am I"' is not assignable to type 'SizeType'.ts(2322)
    render(<Button size={size} />);

    // toHaveBeenCalledWith:
    //   This assertion checks if a function was called at least once with certain parameters.
    //   Requires a spy function to be passed to expect.
    expect(mockWarn).toHaveBeenCalledWith('Warning: [Button] Invalid prop `size`.');

    // mockRestore:
    //   Does what mockReset does and restores inner implementation to the original function.
    //   Note that restoring mock from vi.fn() will set implementation to an empty function that returns undefined.
    //   Restoring a vi.fn(impl) will restore implementation to impl.
    //   If you want this method to be called before each test automatically, you can enable restoreMocks setting in config.
    mockWarn.mockRestore();
  });

  it('should render empty button without errors', () => {
    const { container } = render(
      <Button>
        {null}
        {undefined}
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should change loading state instantly by default', () => {
    const LoadingButton = () => {
      const [loading, setLoading] = useState<ButtonProps['loading']>(false);
      return (
        <Button loading={loading} onClick={() => setLoading(true)}>
          Button
        </Button>
      );
    };
    const { container } = render(<LoadingButton />);
    fireEvent.click(container.firstChild!);
    expect(container.querySelectorAll('.btn-loading').length).toBe(1);
  });

  it('should change loading state with delay', () => {
    const LoadingButton = () => {
      const [loading, setLoading] = useState<ButtonProps['loading']>(false);
      return (
        <Button loading={loading} onClick={() => setLoading({ delay: 1000 })}>
          Button
        </Button>
      );
    };
    const { container } = render(<LoadingButton />);
    fireEvent.click(container.firstChild!);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(0);
  });

  it('reset when loading back of delay', () => {
    // useFakeTimers useRealTimers 配套使用
    vi.useFakeTimers();
    const { rerender, container } = render(<Button loading={{ delay: 1000 }} />);
    rerender(<Button loading={{ delay: 2000 }} />);
    rerender(<Button loading={false} />);

    // act
    //   All it does is forward all arguments to the act function if your version of react supports act.
    //   It is recommended to use the import from @testing-library/react over react-dom/test-utils for consistency reasons.
    act(() => {
      // This method will invoke every initiated timer until the timers queue is empty.
      // It means that every timer called during runAllTimers will be fired.
      // If you have an infinite interval, it will throw after 10 000 tries.
      vi.runAllTimers();
    });

    expect(container.querySelectorAll('.btn-loading')).toHaveLength(0);

    vi.useRealTimers();
  });

  it('should not clickable when button is loading', () => {
    const onClick = vi.fn();
    const {
      container: { firstChild: buttonNode },
    } = render(
      <Button loading onClick={onClick}>
        button
      </Button>
    );
    fireEvent.click(buttonNode!);
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should support link button', () => {
    const { container } = render(
      <Button target="_blank" href="https://ant.design">
        link button
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders {0}, 0 and {false}', () => {
    expect(render(<Button>{0}</Button>).container.firstChild).toMatchSnapshot();
    expect(render(<Button>0</Button>).container.firstChild).toMatchSnapshot();
    expect(render(<Button>{false}</Button>).container.firstChild).toMatchSnapshot();
  });

  it('should not render as link button when href is undefined', async () => {
    const {
      container: { firstChild },
    } = render(
      <Button type="primary" href={undefined}>
        button
      </Button>
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should support to change loading', async () => {
    vi.useFakeTimers();
    const { container, rerender, unmount } = render(<Button>Button</Button>);
    rerender(<Button loading />);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(1);
    rerender(<Button loading={false} />);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(0);
    rerender(<Button loading={{ delay: 50 }} />);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(0);
    await waitFakeTimer(40, 1);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(0);
    await waitFakeTimer(50, 1);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(1);
    rerender(<Button loading={false} />);
    expect(container.querySelectorAll('.btn-loading')).toHaveLength(0);
    expect(unmount).not.toThrow();
    vi.useRealTimers();
  });

  it('should warning when pass type=link and ghost=true', () => {
    resetWarned();
    const warnSpy = vi.spyOn(console, 'error').mockImplementation(() => {});
    render(<Button type="link" ghost />);
    expect(warnSpy).toHaveBeenCalledWith("Warning: [Button] `link` or `text` button can't be a `ghost` button.");
    warnSpy.mockRestore();
  });

  it('should warning when pass type=text and ghost=true', () => {
    resetWarned();
    const warnSpy = vi.spyOn(console, 'error').mockImplementation(() => {});
    render(<Button type="text" ghost />);
    expect(warnSpy).toHaveBeenCalledWith("Warning: [Button] `link` or `text` button can't be a `ghost` button.");
    warnSpy.mockRestore();
  });

  it('should not redirect when button is disabled', () => {
    const onClick = vi.fn();
    const {
      container: { firstChild },
    } = render(
      <Button href="https://ant.design" onClick={onClick} disabled>
        click me
      </Button>
    );
    fireEvent.click(firstChild!);
    expect(onClick).not.toHaveBeenCalled();
  });
});
