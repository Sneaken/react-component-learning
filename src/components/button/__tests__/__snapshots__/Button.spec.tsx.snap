// Vitest Snapshot v1

exports[`Button > renders {0}, 0 and {false} 1`] = `
<button
  class="btn btn-default btn-middle"
  type="button"
>
  0
</button>
`;

exports[`Button > renders {0}, 0 and {false} 2`] = `
<button
  class="btn btn-default btn-middle"
  type="button"
>
  0
</button>
`;

exports[`Button > renders {0}, 0 and {false} 3`] = `
<button
  class="btn btn-default btn-middle"
  type="button"
/>
`;

exports[`Button > renders correctly 1`] = `
<button
  class="btn btn-default btn-middle"
  type="button"
>
  Follow
</button>
`;

exports[`Button > should not render as link button when href is undefined 1`] = `
<button
  class="btn btn-middle btn-primary"
  type="button"
>
  button
</button>
`;

exports[`Button > should render empty button without errors 1`] = `
<button
  class="btn btn-default btn-middle"
  type="button"
/>
`;

exports[`Button > should support link button 1`] = `
<a
  class="btn btn-default btn-middle"
  href="https://ant.design"
  target="_blank"
>
  link button
</a>
`;
